Develop feature mujoco unity

Sohojoe opened the pull request:
I'm getting close with migrating my https://github.com/Sohojoe/MujocoUnity repro into a ml-agent envirment and wanted to get feedback incase I'm off base on the structure. 

I have Humanoid working pretty well and have 3 different trained models at 1, 3, 5, and 8m steps respectively (* 16 agents, so 8m is should be equivilant to 128m physics steps / 25.6 training steps when comparing to other benchmarks)

Short term to do (should be able to complete this in a week or so)
- [x] Finish refactor (removing unused code and make nicer multiple enviroments
- [x] Add OpenAIWalker
- [x] Add OpenAIHopper
- [x] Add OpenAIAnt

Longer term
- [x] improve training speed
- [x] support mujoco nested classes: this will enable DeepmindHumanoid + others

Cut for this release
- reduce observations (we should be able to get down to two observations per joint, I'm having trouble getting accurate local rotation in joint space)
- implement geom axis-angle: this will enable OpenAIHalfCheetah

vincentpierre commented:
Hi @Sohojoe,

This is an awesome PR you made here and the whole team is impressed. I played with some of the environments and every thing went well. 👍 

I have some high level comments. I know you are still working on this, so you are probably aware of most of what follows:

 * There are a lot of warnings when opening the project. I think some of these warnings are about unused variables. If the warnings are non removable, I think you can use `#pragma warning disable` to hide them. 

 * There are a lot of new scripts in `MujocoShared/Scripts`, I think some of them would benefit from a description. You could also include a readme in the folder.

 * `OpenAIHopper/Scripts/ OpenAIHopperAcademy.cs`, `*/New Terrain` contain a space in their names.

 * Some methods have inconsistent naming (for instance `StepReward_Walker101`) `_` and numbers should not be in the method name

If you need help from the team or if there is a feature you need implemented to help you work, let us know.
(It is possible the OpenAIAnt environment behaves like this because of the difference between the Mujoco and the Unity Physics.)

Sohojoe commented:
Hi @vincentpierre, 

That is good feedback. I fixed the warnings, the space in file names, and removed the underscore. 

Regarding the numbers in the reward function name... I do this as way of versioning the reward function + hyperparms with training / trained models. So StepRewardOaiHumanoidRun162() corrilates with the models named humanoid162xxx.bytes. I can remove this once I've wrapped everything up if you would like, but I find this usefull as typically I'll do my heavy lifting on coding at the weekend and then be running a bunch of training during the week while focusing on the day job.

I agree that making a readme / documentation is needed. I'm working on this!

I think the problem with OpenAIAnt is in my MujocoSpawner code. I was able to get some more DeepMind xmls running this weekend which highlighted initialization issues in my MujocoSpawner code. I'm implementing a test bed enviroment to make this easier to debug.

awjuliani commented:
Hi @Sohojoe 

Thanks for taking the time to work on this. I think the community will be really excited to get their hands on these tasks. I am back from vacation this week, and will finally get a chance to look over it properly. 

One general question I have is around the name of the environments. Do we want to call them "Mujoco" tasks, even though there isn't actually mujoco physics being used? Maybe something like "Continuous Control suite" might be better.

Sohojoe commented:
Hi @awjuliani, sounds good! I have no issue changing the name and "Continuous Control suite" works for me.

brancoliticus commented:
Look at the magnificent stupidity of a license that the gym project has: https://github.com/openai/gym/blob/master/LICENSE.md. That is all down to including the xml model files created for Mujoco. As far as I know those are proprietary files that require a license to be bought so as to even download them to a local hard drive. As a consequence OpenAI decided to just drop the ball on top of anyone that might want to use the gym repository for proprietary stuff, you know, like a video game, which is what Unity Technologies specializes in.

Being interested in doing stuff to sell to consumers in the Unity game engine, I wouldn't feel safe to even download this plugin to my hard drive if this pull request were to merge with the Mujoco xml files included, except if I were willing to buy a Mujoco physics engine license for all and I mean all of the users of my eventual video game, which I ain't willing to do in a million years.

Therefore please talk to a company lawyer before merging this stupidity (legal I mean, not technical, technically it may be the best pull request ever created for all I care) in, see what someone that has a professional background in software licensing has to tell about merging third party created files intended to be used with their proprietary software package.

awjuliani commented:
Thanks for brining this up @brancoliticus. We will look into licensing around these files.

Sohojoe commented:
@awjuliani 

I moved the shared code to ML-Agents\Add-ons\ContinuousControl. This is only 3.6mb. I think it good to separate this code from the examples folder so one does not need to clone the example folders to use it.

I renamed the example folder to ML-Agents\Examples\ContinuousControl. The examples take 6.2mb.

The gifs in the docs\images folder are 125.9mb. I could move these to Azure blob storage if you want to reduce the size.

I merged in your latest develop and updated/tested each example.

I renamed 'namespace MojocoUnity' to ‘namespace MLAgents'

Once you decide on a name I will refactor the class names and update the documentation.

awjuliani commented:
Thanks @Sohojoe! I think moving the gifs to cloud storage will be the right call. @vincentpierre and I will take a look at it over the next couple days.
